app/services/record_declarations/base.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+require "json_schema/validate_body_against_schema"
+
+module RecordDeclarations
+  class Base
+    attr_accessor :params
+    delegate :user_profile, :actual_lead_provider, to: :not_implemented_error
+
+    class << self
+      def call(params)
+        new(params).call
+      end
+
+      def required_params
+        %i[participant_id cpd_lead_provider declaration_type declaration_date course_identifier raw_event]
+      end
+    end
+
+    def not_implemented_error
+      raise NotImplementedError, "Method must be implemented"
+    end
+
+    def call
+      validate_schema!
+      add_participant_profile_params!
+
+      declaration = create_record!
+      validate_provider!
+      { id: declaration.id }
+    end
+
+  private
+
+    delegate :school, :cohort, to: :user_profile
+
+    def initialize(params)
+      @params = params
+      @params[:user_id] = params[:participant_id]
+    end
+
+    def validate_schema!
+      errors = JsonSchema::ValidateBodyAgainstSchema.call(schema: schema, body: params[:raw_event])
```

Comment:
Are we sure this is the right place to validate `raw_event`? `ParticipantDeclarationsController` does `validate_params!`, would we not want to validate the raw event there?

Reply:
There are several schemas that we want to validate against. Putting that into the controller would require much more logic than is required with the factory approach which only validates against the schema deemed as valid for that course and type.